### ISSUE TITLE
MPRIS TrackList Interface

### ISSUE THREAD

**Issue by JasonLG1979** [pithos/pithos#425]
I'm going to try my hand at adding support for the MPRIS TrackList Interface to https://github.com/eonpatapon/gnome-shell-extensions-mediaplayer some time in the near future.

I'd also like to do the same for Pithos. I think it would be nice a addition. It would be really neat IMHO to have a "complete" implementation of the MPRIS spec.

The spec is pretty straight forward and it doesn't look like it will that difficult to implement on either side.

The spec mentions limiting the list 20 tracks, but since Pithos can't replay songs anyway my thought is that the list should only contain up coming songs as previous songs offer no real "context" as the spec calls it because we're a radio app, and the context is created by the user as they set the seed for the station. And to me showing the current playing song in the tracklist is redundant.

Your thoughts @TingPing

**Comment by TingPing** [comment]
Sounds good.

**Comment by JasonLG1979** [comment]
@TingPing what do you think about this:
https://specifications.freedesktop.org/mpris-spec/latest/Track_List_Interface.html#Property:Tracks

It's kinda weird that this is the only MPRIS prop that doesn't send a new value. Reading it, it doesn't seem worth even emitting prop changes? Even if I wanted to emit a prop with no value how would I do that? Seems like you're bindings expect a value to be sent?

**Comment by JasonLG1979** [comment]
Well I think I've got the tracklist implemented correctly? It's hard to tell. I'm unaware of any applets/clients that support the tracklist? Know any?
https://github.com/JasonLG1979/pithos/tree/MPRIS-TrackList-Interface

**Comment by TingPing** [comment]
No idea.

**Comment by JasonLG1979** [comment]
Getting there with the extension. Just need to clean up some stuff.

![screenshot from 2017-02-18 03-36-28](https://cloud.githubusercontent.com/assets/6667703/23092251/3db54f12-f58c-11e6-842d-601bcf22de50.png)

**Comment by JasonLG1979** [comment]
OK got it all wired up.
![screenshot from 2017-02-18 11-15-53](https://cloud.githubusercontent.com/assets/6667703/23095234/2b0fe698-f5cc-11e6-998c-d5be454c5e8e.png)

I changed my mind about showing the current song. It's just easier to use the TrackListReplaced signal and replace the whole playlist in one shot as opposed to a series of  TrackAdded and TrackRemoved'  signals. The tracklist title is Pithos specific. Every other player would get "Tracks".

**Comment by JasonLG1979** [comment]
Well here's my latest iteration:
![screenshot from 2017-02-20 00-49-07](https://cloud.githubusercontent.com/assets/6667703/23114897/407defb6-f707-11e6-95e9-5317898674ea.png)

The tracklist logic in the extension much like the playlist logic is pretty simple and idiot proof. Any time you call a method or emit a signal on the interface it just refreshes the entire tracklist. I could make it more complicated but it really doesn't need to be.

You click on the song and it skips to that song. In the case of Pithos only if it's a playable song.

The little dot marks the current song in the tracklist.

The "Current Playlist" title is Pithos specific. All other players will get "Tracks" as a title unless someone puts in a PR to add a custom title for their player much like the Playlists where we get "Stations" and everyone else gets "Playlists".

So far of the dozens of players I've tested Pithos is the only player besides GNOME Music to actually implement the tracklist interface. But GNOME Music's implementation is so horribly broken that it actually causes the player to crash when you call the GoTo method on the interface. I get the same traceback as this bug report: https://bugzilla.redhat.com/show_bug.cgi?id=1236484

I Blacklisted GNOME Music for that interface for now.

The behaviour in Pithos is this:

New playlists replace old playlist in the tracklist when we get a new playlist from Pandora, making sure to always include the current song.

Songs that are banned or tired are removed from the tracklist.

Songs that have been unrated or loved after being previously banned are added back into the tracklist.

Songs that have been tired are never added back into the tracklist as they will be skipped regardless of their new rating.

@TingPing give these to branches a try together when you get a chance and let me know what you think:

https://github.com/JasonLG1979/pithos/tree/MPRIS-TrackList-Interface
https://github.com/JasonLG1979/gnome-shell-extensions-mediaplayer/tree/MPRIS-TrackList-Interface

**Comment by JasonLG1979** [comment]
Tracklist support has been merged into the extension, and works rather well to toot my own horn,lol!!! Just waiting on https://github.com/pithos/pithos/pull/431 @TingPing.

**Comment by JasonLG1979** [comment]
Implemented. Pithos is now the only player that I know of that has a complete (and fully functional) MPRIS implementation. :)
